Publish MinIO events via AMQP: MinIO server stop when RabbitMQ restart

Publish MinIO events via AMQP: minio server stop when rabbitmq restart

## Expected Behavior
minio server normal working when rabbitmq restart

## Current Behavior
minio server stop when rabbitmq restart


## Steps to Reproduce (for bugs)

1. stop rabbitmq and post 2 object. find 2 event in queueDir `/home/events`
```
/home/events/minio-amqp-1 # ls
193dfbc1-749a-4732-b85c-559069b1a831.event  24f13b54-4421-4e80-a214-e87f8ee2097c.event
```
​
2. docker restart rabbitmq, minio server stop, find logs in minio server
​
```
panic: target.Send() failed with 'Exception (501) Reason: "read tcp 172.17.0.2:53246->10.1.10.77:5672: read: connection reset by peer"'
​
goroutine 116 [running]:
github.com/minio/minio/pkg/event/target.sendEvents.func1(0xc00025e0ba, 0x24, 0x2)
	/q/.q/sources/gopath/src/github.com/minio/minio/pkg/event/target/store.go:121 +0x34b
github.com/minio/minio/pkg/event/target.sendEvents(0x1bca360, 0xc0003d05a0, 0xc00047e2a0, 0xc0000a8ae0)
	/q/.q/sources/gopath/src/github.com/minio/minio/pkg/event/target/store.go:142 +0x180
created by github.com/minio/minio/pkg/event/target.NewAMQPTarget
	/q/.q/sources/gopath/src/github.com/minio/minio/pkg/event/target/amqp.go:250 +0x2b3
​
```
​
## Your Environment
<!--- Include as many relevant details about the environment you experienced the bug in -->
* Version used: 
```
/ # minio version
Version: 2019-08-29T00:25:01Z
Release-Tag: RELEASE.2019-08-29T00-25-01Z
Commit-ID: eb18c82976251d9a0c0730df27e10204c89beae2
```
 - minio event config

```
"1": {
    "autoDeleted": false,
    "deliveryMode": 2,
    "durable": true,
    "enable": true,
    "exchange": "bucketevents",
    "exchangeType": "fanout",
    "immediate": false,
    "internal": false,
    "mandatory": false,
    "noWait": false,
    "queueDir": "/home/events",
    "queueLimit": 10000,
    "routingKey": "bucketlogs",
    "url": "amqp://root:passw0rd@10.1.10.77:5672"
   }
```
* Environment name and version :
  - rabbitmq version: 3.7.17
  ```
  docker run -d --hostname my-rabbit --name some-rabbit -e RABBITMQ_DEFAULT_USER=root -e RABBITMQ_DEFAULT_PASS=passw0rd -p 15672:15672 -p 5672:5672 rabbitmq:3-management
  ```
* Operating System and version
```
Linux gx 4.15.0-58-generic #64~16.04.1-Ubuntu SMP Wed Aug 7 14:10:35 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
```

--- a/pkg/event/target/amqp.go
+++ b/pkg/event/target/amqp.go
@@ -231,7 +231,7 @@ func NewAMQPTarget(id string, args AMQPArgs, doneCh <-chan struct{}) (*AMQPTarge
 
 	conn, err = amqp.Dial(args.URL.String())
 	if err != nil {
-		if store == nil || !IsConnRefusedErr(err) {
+		if store == nil || !(IsConnRefusedErr(err) || IsConnResetErr(err)) {
 			return nil, err
 		}
 	}

--- a/pkg/event/target/mysql.go
+++ b/pkg/event/target/mysql.go
@@ -351,7 +351,7 @@ func NewMySQLTarget(id string, args MySQLArgs, doneCh <-chan struct{}) (*MySQLTa
 
 	err = target.db.Ping()
 	if err != nil {
-		if target.store == nil || !IsConnRefusedErr(err) {
+		if target.store == nil || !(IsConnRefusedErr(err) || IsConnResetErr(err)) {
 			return nil, err
 		}
 	} else {

--- a/pkg/event/target/nsq.go
+++ b/pkg/event/target/nsq.go
@@ -182,7 +182,7 @@ func NewNSQTarget(id string, args NSQArgs, doneCh <-chan struct{}) (*NSQTarget, 
 
 	if err := target.producer.Ping(); err != nil {
 		// To treat "connection refused" errors as errNotConnected.
-		if target.store == nil || !IsConnRefusedErr(err) {
+		if target.store == nil || !(IsConnRefusedErr(err) || IsConnResetErr(err)) {
 			return nil, err
 		}
 	}

--- a/pkg/event/target/postgresql.go
+++ b/pkg/event/target/postgresql.go
@@ -358,7 +358,7 @@ func NewPostgreSQLTarget(id string, args PostgreSQLArgs, doneCh <-chan struct{})
 
 	err = target.db.Ping()
 	if err != nil {
-		if target.store == nil || !IsConnRefusedErr(err) {
+		if target.store == nil || !(IsConnRefusedErr(err) || IsConnResetErr(err)) {
 			return nil, err
 		}
 	} else {

--- a/pkg/event/target/redis.go
+++ b/pkg/event/target/redis.go
@@ -276,7 +276,7 @@ func NewRedisTarget(id string, args RedisArgs, doneCh <-chan struct{}) (*RedisTa
 
 	_, pingErr := conn.Do("PING")
 	if pingErr != nil {
-		if target.store == nil || !IsConnRefusedErr(pingErr) {
+		if target.store == nil || !(IsConnRefusedErr(pingErr) || IsConnResetErr(pingErr)) {
 			return nil, pingErr
 		}
 	} else {

--- a/pkg/event/target/store.go
+++ b/pkg/event/target/store.go
@@ -93,8 +93,12 @@ func IsConnRefusedErr(err error) bool {
 	return false
 }
 
-// isConnResetErr - Checks for connection reset errors.
-func isConnResetErr(err error) bool {
+// IsConnResetErr - Checks for connection reset errors.
+func IsConnResetErr(err error) bool {
+	if strings.Contains(err.Error(), "connection reset by peer") {
+		return true
+	}
+	// incase if error message is wrapped.
 	if opErr, ok := err.(*net.OpError); ok {
 		if syscallErr, ok := opErr.Err.(*os.SyscallError); ok {
 			if syscallErr.Err == syscall.ECONNRESET {
@@ -117,7 +121,7 @@ func sendEvents(target event.Target, eventKeyCh <-chan string, doneCh <-chan str
 				break
 			}
 
-			if err != errNotConnected && !isConnResetErr(err) {
+			if err != errNotConnected && !IsConnResetErr(err) {
 				panic(fmt.Errorf("target.Send() failed with '%v'", err))
 			}
 

